analyzers/src/SonarAnalyzer.CSharp/Rules/ThrowReservedExceptions.cs

```diff
@@ -18,23 +18,19 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-namespace SonarAnalyzer.Rules.CSharp
-{
-    [DiagnosticAnalyzer(LanguageNames.CSharp)]
-    public sealed class ThrowReservedExceptions : ThrowReservedExceptionsBase
-    {
-        private static readonly DiagnosticDescriptor rule =
-            DescriptorFactory.Create(DiagnosticId, MessageFormat);
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+namespace SonarAnalyzer.Rules.CSharp;
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
-            context.RegisterNodeAction(
-                c => ReportReservedExceptionCreation(c, ((ThrowStatementSyntax)c.Node).Expression),
-                SyntaxKind.ThrowStatement);
-        }
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ThrowReservedExceptions : ThrowReservedExceptionsBase<SyntaxKind>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
 
-        protected override bool IsObjectCreation(SyntaxNode throwStatementExpression) =>
-            throwStatementExpression.IsKind(SyntaxKind.ObjectCreationExpression);
+    protected override void Initialize(SonarAnalysisContext context)
+    {
+        context.RegisterNodeAction(c => Process(c, ((ThrowStatementSyntax)c.Node).Expression), SyntaxKind.ThrowStatement);
+        context.RegisterNodeAction(c => Process(c, ((ThrowExpressionSyntaxWrapper)c.Node).Expression), SyntaxKindEx.ThrowExpression);
```

Comment:
Open question (and wild suggestion): It's not the first rule that raises for both SyntaxKindEx.ThrowExpression and SyntaxKind.ThrowStatement, I was wondering if we could add an `abstract TSyntaxKind[] ThrowSyntaxKinds { get; }` in Language.SyntaxKind, to centralize the initializer in the base class (just by adding an abstract `GetExpression` method with a switch to return the SyntaxNode to be fed to `Process`.
It might be useful also for: 
- DoNotThrowFromDestructors
- RedundantJumpStatement
- TrackNotImplementedException

Reply:
Only the 1st one has VB version and depends on factory currently. 
But the idea is valid, if you extend it by making the ThrowExpression a syntax facade method.

Since its used only twice so far, I won't do it. I'd do it for three according to "Two is a group, three is a crowd" rule

Reply:
Fine for me! BTW I think there was also this one :smile:

> 0 is none, 1 is single, 2 is many.